cpp/serve/prefix_cache.cc

```diff
@@ -0,0 +1,260 @@
+   */
+  void RollBackSequence(int64_t seq_id, size_t num_tokens) {
+    ++lru_counter;
+    radix_tree->RollBackSequence(seq_id, num_tokens);
+    latest_visit[seq_id] = lru_counter;
+  }
+
+  /*!
+   * \brief Recycle a sequence. The recycled sequence will not be removed immediately, as long as
+   memory is sufficient. And it will be reused again in the future request.
+   * \param seq_id The sequence to be recycled.
+   * \param callback The callback function to be invoked when removing the sequence.
+   * \param lazy The flag if the sequence should be removed lazily or intermediary.
+   * \throw Error if the given sequence id is not valid.
+   */
+  void RecycleSequence(int64_t seq_id, PackedFunc callback, bool lazy = true) {
+    CHECK(latest_visit.find(seq_id) != latest_visit.end());
+    size_t timestamp = latest_visit[seq_id];
+    CHECK(recycling_seqs.find({timestamp, seq_id}) == recycling_seqs.end());
+    CHECK(recycle_callbacks.find(seq_id) == recycle_callbacks.end());
+    if (lazy) {
+      // Remove the sequence lazily.
+      if (recycle_callbacks.size() == max_recycling_seqs) {
+        // If the number of recycling sequence has reached the maximum values, pop the oldest one.
+        TryFreeMemory();
+      }
+      recycle_callbacks[seq_id] = callback;
+      recycling_seqs.emplace(latest_visit[seq_id], seq_id);
+    } else {
+      // Remove the sequence intermediately.
+      radix_tree->RemoveSequence(seq_id);
+      callback();
+    }
+  }
+
+  /*!
+   * \brief Try to remove recycling sequence to free up memory. It will remove the oldest recycling
+   sequence.
+   * \return The flag if there is a sequence removed. In other word, return true when memory is
+   freed successfully.
+   * \throw Error if the given sequence id is not valid.
+   */
+  bool TryFreeMemory() {
+    if (recycling_seqs.empty()) {
+      // There is no recycling sequence. No memory can be freed.
+      return false;
+    }
+    auto it = recycling_seqs.begin();
+    int64_t seq_id = it->second;
+    CHECK(recycle_callbacks.find(seq_id) != recycle_callbacks.end());
+    CHECK(latest_visit.find(seq_id) != latest_visit.end());
+    radix_tree->RemoveSequence(seq_id);
+    recycle_callbacks[seq_id]();
+
+    recycle_callbacks.erase(seq_id);
+    recycling_seqs.erase(it);
+    latest_visit.erase(seq_id);
+    return true;
+  }
+
+  /*!
+   * \brief Check if a sequence exists.
+   * \param seq_id The sequence ID for index.
+   * \return The sequence existence.
+   * \throw Error if sequence ID is not valid.
+   */
+  bool HasSequence(int64_t seq_id) { return radix_tree->HasSequence(seq_id); }
+
+  /*!
+   * \brief Reset the prefix cache to initial status.
+   */
+  void Reset() {
+    radix_tree->Reset();
+    latest_visit.clear();
+    recycling_seqs.clear();
+    recycle_callbacks.clear();
+    lru_counter = 0;
+  }
+
+ private:
+  /*!
+   * \brief The core data structure radix tree.
+   */
+  PagedRadixTree radix_tree;
```

Comment:
Code style: class member must ends with underscore, `PagedRadixTree radix_tree_`, this is for better readability